### PR TITLE
Fix issue when backward graph returning sharded tensors

### PR DESCRIPTION
**Pull request by aazzolini**
When SPMD is part of a larger model, backard() could output sharded
    gradients (for input) as well as replicated gradients (for parameters).
    
    This fixes a bug that was causing for non-partial/ non-replicated
    outputs to disappear from the graph output.
    
    --
    
    Also, not necessary for this specific fix, but something that could
    trigger issues down the road -- the output_specs were not reflecring the
    Partial -> Replicated  resharding on the outputs. This fixes it so that
    _convert_to_distributed always returns the proper output sharding for
    the graph.
    (One thing we can do since we have the output schemas: return DTensors from SPMD (optionally)

### CODE REVIEW

File: spmd/compiler/distribute.py [hunk 1]
```diff
@@ -165,22 +165,17 @@ def remap_arg(arg: object) -> object:
     def unwrap_local(e: object) -> object:
         return e._local_tensor if isinstance(e, DTensor) else e
 
-    gm = make_fx(dispatch)(local_target_args)
-    for nd in gm.graph.nodes:
-        if nd.name == "_tensor_constant0":
-            print(
-                f"WARNING: found constant while SPMD expansion for {node}. Graph {gm}. This is not supported."
-            )
-    return gm
```
Comment (aazzolini) [inline spmd/compiler/distribute.py:174]: I realized this warning is wrong, i think a constnat is always expected (It's the ProcessGroup i think).

File: spmd/compiler/distribute.py [hunk 2]
```diff
@@ -206,38 +201,44 @@ def dummy_add(grad: torch.Tensor, zero: torch.Tensor) -> torch.Tensor:
...
             continue
-        obj = node_to_obj[argument]
 
-        has_partial = _is_partial_dtensor(obj)
+        obj = node_to_obj[argument]
 
-        if not has_partial:
+        if not _is_partial_dtensor(obj):
+            new_args.append(argument)
             continue
```
Comment (aazzolini) [inline spmd/compiler/distribute.py:223]: This is the main fix . There's 2 issues

1) has_partial was set to True and then back to False if a Sharded() output appeared after a Partial_() output (I think this bug was added in the cleanup).

2) new_args needs to be appended with the output even if it's not partial. Thsi was being skipped -- and this is what caused less outputs than expected.

**Review by richqyz — APPROVED** [review]
Thanks for fixing this!

File: spmd/compiler/distribute.py (hunk 2 above)
Comment (anj-s) [reply spmd/compiler/distribute.py:223]: "new_args needs to be appended with the output even if it's not partial." - Can you explain this a bit more? Not sure I got this issue.

**Review by anj-s — APPROVED** [review]

File: spmd/compiler/distribute.py (hunk 2 above)
Comment (aazzolini) [reply spmd/compiler/distribute.py:223]: In this loop we are iterating through all the outputs.
We need to check, for each output, if they are Partial or not. 
- If they are Partial, we add the all reduce, and then append the output from all_reduce to the output list
- If they are not Partial, we should just append them to the new output list. But we were not doing that, so here's the bug.

**Pull request merged**